src/core/qgsmaprendererjob.cpp

```diff
@@ -304,9 +298,23 @@ LayerRenderJobs QgsMapRendererJob::prepareJobs( QPainter *painter, QgsLabelingEn
     if ( mFeatureFilterProvider )
       job.context.setFeatureFilterProvider( mFeatureFilterProvider );
 
+    bool hasStyleOverride = mSettings.layerStyleOverrides().contains( ml->id() );
+    if ( hasStyleOverride )
+      ml->styleManager()->setOverrideStyle( mSettings.layerStyleOverrides().value( ml->id() ) );
+
+    job.blendMode = ml->blendMode();
+    job.opacity = 1.0;
+    if ( QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( ml ) )
+    {
+      job.opacity = vl->opacity();
+    }
+
     // if we can use the cache, let's do it and avoid rendering!
     if ( mCache && mCache->hasCacheImage( ml->id() ) )
     {
+      if ( hasStyleOverride )
```

Comment:
I think this block is calling out for a small RAII class to handle the style override restore - something like ExpressionContextScopePopper

Reply:
@nyalldawson , will do.